README.md
# GStremer Boilerplate Element Plugin with CMake Build System

# Building

Navigate to the top project directory, same directory this readme is in.

Make a build director and enter that directory

    mkdir build
    cd build

Setup the cmake project

    # the '..' tells cmake to go back one dir to find the CMakeLists txt file
    cmake ..

Finally build the plugin

    make

# Testing

Tell gstreamer where to look for the newly build plugin

    # run from the build dir
    export GST_PLUGIN_PATH_1_0=$PWD

Use `gst-inspect-1.0` to check the new plugin

    gst-inspect-1.0 boilerplatevideofilter

Test the element in a pipeline

    gst-launch-1.0 videotestsrc ! boilerplatevideofilter ! videoconvert ! fpsdisplaysink

# Generating a Debian Package

Current a binary debian package is used for releasing the gst_boilerplatevideofilter plugin

To build a package

    # Follow the build instructions above

Run make debgen to generate the debian packages

    make debgen

The result will be a debian package in the build directory called 'libgstboilerplatevideofilter-0.1.0.0-aarch64-main-release.deb'

# Installing debian Package

To install this package:

    sudo dpkg -i relative_path_to_deb_file

Use gst-inspect-1.0 to check if the package was installed correctly

    gst-inspect-1.0 boilerplatevideofilter

Note:

The standard gstreamer plugin install path is `/usr/lib/x86_64-linux-gnu//gstreamer-1.0` you may want to install in /opt and tell gstreamer to look there for more plugins

# Installing the element on x86 Ubuntu

Follow just the bild instructions above then tell gstreamer where to fin the element binary.

For example:

    export GST_PLUGIN_PATH_1_0=$PWD


## Removing installed .deb packages

    sudo apt-get remove gst-boilerplatevideofilter

name_of_package is the name of the package in the control file, not the name of the .deb

It may also be helpful to check the installed version is as you expect

    apt-cache policy name_of_package

# Helpful GStreamer stuff

    # check for blacklisted pugins
    gst-inspect-1.0 -b

    # clear blacklister plugins
    rm ~/.cache/gstreamer-1.0/registry*

    find . -name "foo*"

    # enable gstreamer debugging
    export GST_DEBUG=4

# Helpful resources

## Good Tutorial for GTK signals

https://www.cc.gatech.edu/data_files/public/doc/gtk/tutorial/gtk_tut-2.html

## See post here for explenation of gstreamer element communication methods

http://gstreamer-devel.966125.n4.nabble.com/Pass-extra-information-between-elements-in-a-pipeline-td4669768.html


CHANGELOG.md
# GStreamer ObjectTrack Plugin Change Log

## 0.1.0.20 / 16 June 2020

- Fixed bug causing missing y rate scalar property

## 0.1.0.19 / 16 June 2020

- Increase maximum x and y deadband from 50 to 200

## 0.1.0.18 / 2 June 2020

- Fix fliped yaw and pitch rate scalars

## 0.1.0.17 / 25 May 2020

- Flip sign of default yaw rate scalar

## 0.1.0.16 / 19 Mar 2020

- Initial release

## 0.1.0.0 / UNRELEASED

- Scene track functionality reached

test/objectTrack_test.h
//
//  objecttrack_test.h
//
//  Created by Isaac Reed on 3/9/20
//

#include <iostream>
#include <signal.h>
#include <gst/gst.h>
#include <gst/app/gstappsink.h>
#include <gst/controller/gstinterpolationcontrolsource.h>
#include <gst/controller/gstdirectcontrolbinding.h> 
#include <string>
#include <sstream>
#include <pthread.h>        
#include <stdlib.h>
#include <termios.h>
#include <unistd.h>
#include <stdio.h>
#include <execinfo.h>

typedef struct _Elements{
  GstElement *pipeline,
             *source1,
             *nvvidconv,
             *videoconvert,
             *source1_queue,
             *objecttrack,
             *sink_queue,
             *sink_videoconvert,
             *sink;
  GstCaps *source1_caps, *videoconvert_caps, *bgr_caps;
  GMainLoop *loop;

}Elements;

Elements elements;

test/objecttrack_test.cpp
//
//  objecttrack_test.h
//
//  Created by Isaac Reed on 3/9/20
//

// gst-launch-1.0 nvarguscamerasrc sensor-mode=0 ! 
// 'video/x-raw(memory:NVMM), width=2592, height=1944' ! 
// nvvidconv ! 'video/x-raw, format=RGBA, width=480, height=360' ! 
// videoconvert ! 'video/x-raw, format=BGR' ! objecttrack ! 
// valve drop=false ! videoconvert ! nveglglessink qos=false window-x=1440
//

#include "objectTrack_test.h"

using namespace std;

bool signal_recieved = false;
int sigint_count = 0;

void sig_handler(int signo){
	if( signo == SIGINT ){
		//printf("received SIGINT\n");
    if(sigint_count == 3){
      cout << "\ni said please enter 'q' to exit" << endl;
    }else if(sigint_count == 6){
      cout << "\nok seriously just enter 'q' it is not that hard, the key next to the 'w'" << endl;
    }else if(sigint_count == 9){
      cout << "\nsorry wrong again enter 'q' to exit" << endl;
    }else if(sigint_count == 12){
      cout << "\nnope, the letter 'q' thats what you want" << endl;
    }else if(sigint_count == 15){
      cout << "\n no no no no, still wrong" << endl;
    }else{
      cout << "\nplease enter 'q' to exit" << endl;
    }
		//signal_recieved = true;
    sigint_count ++;

	}
}

void *TermioControl(void *ptr){
  cout << "termio control running" << endl;
  struct termios t;
  tcgetattr(STDIN_FILENO, &t); //get the current terminal I/O structure
  t.c_lflag &= ~ICANON; //Manipulate the flag bits to do what you want it to do
  tcsetattr(STDIN_FILENO, TCSANOW, &t); //Apply the new settings

  while(!signal_recieved){
    char input;
    cin >> input;
    cout << " | ";
    switch(input){
      case 27:
      case 'q':
        cout << "quit" << endl;
        signal_recieved = true;
        break;

      case 's':
        gboolean enabled;
        cout << "scene track toggle" << endl;
        g_object_get(G_OBJECT(elements.objecttrack), "enable", &enabled, NULL);
        g_object_set(G_OBJECT(elements.objecttrack), "enable", !enabled, NULL);
        break;
      
      case 'p':
        break;
      
      case '=':
      case 't':
        break;

      case '-':
      case 'w':
        break;

      case 92:  /* \ */
        break;

      case '[':
        break;

      case ']':
        break;
      
      case '1':
        cout << "switch to cam 1" << endl;
        break;

      case '2':
        cout << "switch to cam 2" << endl;
        break;

      case '3':
        cout << "switch to cam 3" << endl;
        break;
        
      case '4':
        cout << "switch to cam 4" << endl;
        break;

      case '5':
        cout << "switch to cam 5" << endl;
        break;

      case 'a':
        break;

      default:
        break;
    }
  }
  cout << "termio control exit" << endl;
}

static gboolean
bus_callback (GstBus * bus, GstMessage * message)
{
  const GstStructure *msg_struct;
  //g_print ("Got %s message\n", GST_MESSAGE_TYPE_NAME (message));
  switch (GST_MESSAGE_TYPE (message)) {
    case GST_MESSAGE_ERROR:{
      GError *err;
      gchar *debug;

      gst_message_parse_error (message, &err, &debug);
      g_print ("Error: %s\n", err->message);
      g_error_free (err);
      g_free (debug);

      signal_recieved = true;
      break;
    }
    case GST_MESSAGE_EOS:
      /* end-of-stream */
      signal_recieved = true;
      break;

    case GST_MESSAGE_ELEMENT:
      msg_struct = gst_message_get_structure (message);
        if (gst_structure_has_name (msg_struct, "objecttrack-rate-out")){
          const GValue *pitch_rate_g = gst_structure_get_value(msg_struct, "pitch-rate");
          const GValue *yaw_rate_g = gst_structure_get_value(msg_struct, "yaw-rate");
          float pitch_rate = g_value_get_float(pitch_rate_g);
          float yaw_rate = g_value_get_float(yaw_rate_g);
          // cout << "recieved objecttrack rate out message "
          //      << " pitch " << pitch_rate 
          //      << " yaw " << yaw_rate << endl;
          
        }
      break;


    default:
      /* unhandled message */
      break;
  }

  /* we want to be notified again the next time there is a message
   * on the bus, so returning TRUE (FALSE means we want to stop watching
   * for messages on the bus and our callback should not be called again)
   */
  return TRUE;
}

bool SourceSetup(){
  g_object_set(G_OBJECT(elements.source1), "sensor-id", 0, NULL);
  g_object_set(G_OBJECT(elements.source1), "sensor-mode", 2, NULL);
  g_object_set(G_OBJECT(elements.source1), "maxperf", true, NULL);
  
  // source 1 caps filters
  if(!gst_element_link_filtered(elements.source1, elements.nvvidconv, elements.source1_caps)){
    cout << "source 1 nvvidconv caps link failed" << endl;
    return false;
  }

  if(!gst_element_link_filtered(elements.nvvidconv, elements.source1_queue, elements.videoconvert_caps)){
    cout << "source 1 videoconvert caps link failed" << endl;
    return false;
  }
  return true;
}

bool SinkSetup(){
  g_object_set(G_OBJECT(elements.sink), "qos", false, NULL);
  g_object_set(G_OBJECT(elements.sink), "window-x", 1440, NULL);
  return true;
}

int main(int argc, char *argv[]){

  if(signal(SIGINT, sig_handler) == SIG_ERR){
    cout << "cannot register signal handler" << endl;
  }

  GstBus *bus;
  guint bus_watch_id;
  GstStateChangeReturn ret;
  GMainLoop *main_loop;

  // GStreamer library init 
  gst_init(NULL, NULL);

  elements.pipeline = gst_pipeline_new("pipeline");
  elements.source1 = gst_element_factory_make("nvarguscamerasrc", "source1");
  elements.nvvidconv = gst_element_factory_make("nvvidconv", "capture-nvvidconv");
  elements.videoconvert = gst_element_factory_make("videoconvert", "capture-videoconvert");
  elements.source1_queue = gst_element_factory_make("queue", "source-queue");
  elements.objecttrack = gst_element_factory_make("objecttrack", "objectrack1");
  elements.sink_videoconvert = gst_element_factory_make("videoconvert", "sink-videoconvert");
  elements.sink_queue = gst_element_factory_make("queue", "sink-queue");
  elements.sink = gst_element_factory_make("nveglglessink", "sink");

  // caps filters
  elements.source1_caps = gst_caps_from_string("video/x-raw(memory:NVMM), width=1920, height=1080");
  elements.videoconvert_caps = gst_caps_from_string("video/x-raw, format=RGBA, width=480, height=360");
  elements.bgr_caps = gst_caps_from_string("video/x-raw, format=BGR");
  bool elements_ok = true;

  if(!elements.pipeline){ cerr << "failed to create pipeine" << endl; elements_ok = false;}
  if(!elements.source1){ cerr << "failed to create source1 element" << endl; elements_ok = false;}
  if(!elements.nvvidconv){ cerr << "failed to create nvvidconv element" << endl; elements_ok = false;}
  if(!elements.videoconvert){ cerr << "failed to create videoconvert element" << endl; elements_ok = false;}
  if(!elements.source1_queue){ cerr << "failed to create source1_queue element" << endl; elements_ok = false;}
  if(!elements.objecttrack){ cerr << "failed to create objectrack element" << endl; elements_ok = false;}
  if(!elements.videoconvert){ cerr << "failed to create videoconvert element" << endl; elements_ok = false;}
  if(!elements.sink_queue){ cerr << "failed to create sink queue element" << endl; elements_ok = false;}
  if(!elements.sink){ cerr << "failed to create sink element" << endl; elements_ok = false;}

  if(elements_ok == false){
    cout << "not all elements could be created, exiting" << endl;
    return -1;
  }

  gst_bin_add_many(GST_BIN(elements.pipeline), elements.source1, elements.nvvidconv,
                           elements.videoconvert, elements.source1_queue,
                           elements.objecttrack, elements.sink_queue,
                           elements.sink_videoconvert, elements.sink, NULL);

  if(!SourceSetup()){
    gst_object_unref(elements.pipeline);
    return -1;
  }

  if(!SinkSetup()){
    gst_object_unref(elements.pipeline);
    return -1;
  }

  gst_element_link(elements.source1_queue, elements.videoconvert);

  if(!gst_element_link_filtered(elements.videoconvert, elements.objecttrack, elements.bgr_caps)){
    cout << "bgr caps link failed" << endl;
    gst_object_unref(elements.pipeline);
    return -1;
  }

  gst_element_link_many(elements.objecttrack, elements.sink_videoconvert, elements.sink, NULL);

  bus = gst_pipeline_get_bus(GST_PIPELINE(elements.pipeline));

  ret = gst_element_set_state(elements.pipeline, GST_STATE_PLAYING);
  if (ret == GST_STATE_CHANGE_FAILURE) {
    cout << "Unable to set the pipeline to the playing state." << endl;
    gst_object_unref(elements.pipeline);
    return -1;
  }
  
  pthread_t termio_thread;
    int err = pthread_create(&termio_thread, NULL, &TermioControl, NULL);
  if(err){
    cout << "termio thread creation failed, "
            "user control will not be available" << endl;
  }

  gst_bus_enable_sync_message_emission(bus);
  usleep(500);
  GstMessage *msg;
  while(!signal_recieved){
    usleep(50);
    msg = gst_bus_pop(bus);
    if(msg != NULL){
      bus_callback(bus, msg);
    }
  }
  gst_message_unref(msg);

  err = pthread_join(termio_thread, NULL);
  if(err){
    cout << "termio thread join fail err: " << err << endl;
  }else{
    cout << "termio thread join success" << endl;
  }

  gst_element_set_state(elements.pipeline, GST_STATE_NULL);

  return 0;
}

src/gstboilerplatevideofilter.h
/* GStreamer
 * Copyright (C) 2020 FIXME <ireed@altavian.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifndef _GST_BOILERPLATE_VIDEOFILTER_H_
#define _GST_BOILERPLATE_VIDEOFILTER_H_

#include <gst/video/video.h>
#include <gst/video/gstvideofilter.h>
#include "version.h"

G_BEGIN_DECLS

#define GST_TYPE_BOILERPLATE_VIDEOFILTER   (gst_boilerplatevideofilter_get_type())
#define GST_BOILERPLATE_VIDEOFILTER(obj)   (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_BOILERPLATE_VIDEOFILTER,GstBoilerPlateVideoFilter))
#define GST_BOILERPLATE_VIDEOFILTER_CLASS(klass)   (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_BOILERPLATE_VIDEOFILTER,GstBoilerPlateVideoFilterClass))
#define GST_IS_BOILERPLATE_VIDEOFILTER(obj)   (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_BOILERPLATE_VIDEOFILTER))
#define GST_IS_BOILERPLATE_VIDEOFILTER_CLASS(obj)   (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_BOILERPLATE_VIDEOFILTER))

typedef struct _GstBoilerPlateVideoFilter GstBoilerPlateVideoFilter;
typedef struct _GstBoilerPlateVideoFilterClass GstBoilerPlateVideoFilterClass;

struct _GstBoilerPlateVideoFilter
{
  GstVideoFilter base_boilerplatevideofilter;

  GstPad *srcpad;
  GstPad *ir_sinkpad;
  GstPad *eo_sinkpad;

  GstCollectPads *cpads;

  gint count;

  // properties
  gint int_property;
  bool bool_property;

};

struct _GstBoilerPlateVideoFilterClass
{
  GstVideoFilterClass base_boilerplatevideofilter_class;
};

GType gst_boilerplatevideofilter_get_type (void);

G_END_DECLS

#endif


src/gstboilerplatevideofilter.cpp
/* GStreamer
 * Copyright (C) 2020 Isaac Reed <ireed@altavian.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Suite 500,
 * Boston, MA 02110-1335, USA.
 */
/**
 * SECTION:element-boilerplatevideofilter
 *
 * Boilerplate template video filter element using cmake build system
 *
 * <refsect2>
 * <title>Example launch line</title>
 * |[
 * gst-launch-1.0 videotestsrc ! boilerplatevideofilter ! fakevideosink
 * ]|
 * </refsect2>
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <gst/gst.h>
#include <gst/video/video.h>
#include <gst/video/gstvideofilter.h>
#include <gst/base/gstcollectpads.h>
#include "gstboilerplatevideofilter.h"

GST_DEBUG_CATEGORY_STATIC (gst_boilerplatevideofilter_debug_category);
#define GST_CAT_DEFAULT gst_boilerplatevideofilter_debug_category

static GstStaticPadTemplate eo_sink_factory = GST_STATIC_PAD_TEMPLATE ("eo_sink",
    GST_PAD_SINK,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE ("{ BGR }"))
    );

static GstStaticPadTemplate ir_sink_factory = GST_STATIC_PAD_TEMPLATE ("ir_sink",
    GST_PAD_SINK,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE ("{ BGR }"))
    );

/* prototypes boilerplatevideofilter */
static void gst_boilerplatevideofilter_set_property (GObject * object,
    guint property_id, const GValue * value, GParamSpec * pspec);
static void gst_boilerplatevideofilter_get_property (GObject * object,
    guint property_id, GValue * value, GParamSpec * pspec);
static gboolean gst_boilerplatevideofilter_set_info (GstVideoFilter * filter,
    GstCaps * incaps, GstVideoInfo * in_info, GstCaps * outcaps,
    GstVideoInfo * out_info);

static GstFlowReturn gst_boilerplatevideofilter_transform_frame (GstVideoFilter * filter,
    GstVideoFrame * inframe, GstVideoFrame * outframe);
static GstFlowReturn gst_boilerplatevideofilter_transform_frame_ip (GstVideoFilter *
    filter, GstVideoFrame * frame);

static GstFlowReturn gst_boilerplatevideofilter_collect_pads (GstCollectPads * cpads,
    GstBoilerPlateVideoFilter * boilerplatevideofilter);

#define DEFAULT_INT_PROPERTY 5
#define DEFAULT_BOOL_PROPERTY 0

enum{
  PROP_0,
  PROP_INT_PROPERTY,
  PROP_BOOL_PROPERTY
};

/* pad templates */
#define VIDEO_SRC_CAPS \
    GST_VIDEO_CAPS_MAKE("{ BGR }")

#define VIDEO_SINK_CAPS \
    GST_VIDEO_CAPS_MAKE("{ BGR }")

/* class initialization */

G_DEFINE_TYPE_WITH_CODE (GstBoilerPlateVideoFilter, gst_boilerplatevideofilter, GST_TYPE_VIDEO_FILTER,
    GST_DEBUG_CATEGORY_INIT (gst_boilerplatevideofilter_debug_category, "boilerplatevideofilter", 0,
        "debug category for boilerplatevideofilter element"));

static void
gst_boilerplatevideofilter_class_init (GstBoilerPlateVideoFilterClass * klass)
{
  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
  GstBaseTransformClass *base_transform_class =
      GST_BASE_TRANSFORM_CLASS (klass);
  GstVideoFilterClass *video_filter_class = GST_VIDEO_FILTER_CLASS (klass);

  /* Setting up pads and setting metadata should be moved to
     base_class_init if you intend to subclass this class. */
  gst_element_class_add_pad_template (GST_ELEMENT_CLASS (klass),
      gst_pad_template_new ("src", GST_PAD_SRC, GST_PAD_ALWAYS,
          gst_caps_from_string (VIDEO_SRC_CAPS)));
  // gst_element_class_add_pad_template (GST_ELEMENT_CLASS (klass),
  //     gst_pad_template_new ("sink", GST_PAD_SINK, GST_PAD_ALWAYS,
  //         gst_caps_from_string (VIDEO_SINK_CAPS)));

  gst_element_class_add_static_pad_template (GST_ELEMENT_CLASS (klass), 
      &eo_sink_factory);
  gst_element_class_add_static_pad_template (GST_ELEMENT_CLASS (klass), 
      &ir_sink_factory);
  
  gst_element_class_set_static_metadata (GST_ELEMENT_CLASS (klass),
      "Gstreamer Object Tracker", "Generic", "Object and scene tracking gsreamer plugin",
      "Isaac Reed <ireed@altavian.com>");

  gobject_class->set_property = gst_boilerplatevideofilter_set_property;
  gobject_class->get_property = gst_boilerplatevideofilter_get_property;
  video_filter_class->set_info = GST_DEBUG_FUNCPTR (gst_boilerplatevideofilter_set_info);
  video_filter_class->transform_frame_ip =
      GST_DEBUG_FUNCPTR (gst_boilerplatevideofilter_transform_frame_ip);


  g_object_class_install_property(gobject_class, PROP_INT_PROPERTY,
    g_param_spec_int("int-property", "An example of an integer property",
      "An example of an integer property",
      0, 200, DEFAULT_INT_PROPERTY, 
      (GParamFlags) (G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS)));

  g_object_class_install_property(gobject_class, PROP_BOOL_PROPERTY,
    g_param_spec_boolean("bool-property", "An example of an boolean property",
      "An example of an boolean property",
       DEFAULT_BOOL_PROPERTY, 
      (GParamFlags) (G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS)));
}

static void
gst_boilerplatevideofilter_init (GstBoilerPlateVideoFilter * boilerplatevideofilter)
{
  boilerplatevideofilter->cpads = gst_collect_pads_new ();
  gst_collect_pads_set_function (boilerplatevideofilter->cpads,
      (GstCollectPadsFunction) GST_DEBUG_FUNCPTR (gst_boilerplatevideofilter_collect_pads),
      boilerplatevideofilter);

  boilerplatevideofilter->eo_sinkpad = 
      gst_pad_new_from_static_template(&eo_sink_factory, eo_sink_factory.name_template);
  GST_PAD_SET_PROXY_CAPS(boilerplatevideofilter->eo_sinkpad);
  gst_element_add_pad (GST_ELEMENT (boilerplatevideofilter),
      boilerplatevideofilter->eo_sinkpad);
  
  boilerplatevideofilter->ir_sinkpad = 
      gst_pad_new_from_static_template(&ir_sink_factory, ir_sink_factory.name_template);
  gst_element_add_pad (GST_ELEMENT (boilerplatevideofilter),
      boilerplatevideofilter->ir_sinkpad);

  gst_collect_pads_add_pad (boilerplatevideofilter->cpads, boilerplatevideofilter->eo_sinkpad,
      sizeof (GstCollectData), NULL, TRUE);
  gst_collect_pads_add_pad (boilerplatevideofilter->cpads, boilerplatevideofilter->ir_sinkpad,
      sizeof (GstCollectData), NULL, TRUE);

  // set default prorerty values
  boilerplatevideofilter->int_property = DEFAULT_INT_PROPERTY;
  boilerplatevideofilter->bool_property = DEFAULT_BOOL_PROPERTY;
}

static GstFlowReturn
gst_boilerplatevideofilter_collect_pads (GstCollectPads * cpads, GstBoilerPlateVideoFilter * boilerplatevideofilter)
{
  GstBuffer *buf1, *buf2;
  GstCaps *caps1, *caps2;

  // buf1 = gst_collect_pads_pop (comp->cpads,
  //     gst_pad_get_element_private (comp->sinkpad));
  // caps1 = gst_pad_get_current_caps (comp->sinkpad);

  // buf2 = gst_collect_pads_pop (comp->cpads,
  //     gst_pad_get_element_private (comp->checkpad));
  // caps2 = gst_pad_get_current_caps (comp->checkpad);

  // if (!buf1 && !buf2) {
  //   gst_pad_push_event (comp->srcpad, gst_event_new_eos ());
  //   return GST_FLOW_EOS;
  // } else if (buf1 && buf2) {
  //   gst_compare_buffers (comp, buf1, caps1, buf2, caps2);
  // } else {
  //   GST_WARNING_OBJECT (comp, "buffer %p != NULL", buf1 ? buf1 : buf2);

  //   comp->count++;
  //   gst_element_post_message (GST_ELEMENT (comp),
  //       gst_message_new_element (GST_OBJECT (comp),
  //           gst_structure_new ("delta", "count", G_TYPE_INT, comp->count,
  //               NULL)));
  // }

  // if (buf1)
  //   gst_pad_push (comp->srcpad, buf1);

  // if (buf2)
  //   gst_buffer_unref (buf2);

  // if (caps1)
  //   gst_caps_unref (caps1);

  // if (caps2)
  //   gst_caps_unref (caps2);

  return GST_FLOW_OK;
}

void
gst_boilerplatevideofilter_set_property (GObject * object, guint property_id,
    const GValue * value, GParamSpec * pspec)
{
  GstBoilerPlateVideoFilter *boilerplatevideofilter = GST_BOILERPLATE_VIDEOFILTER (object);

  GST_DEBUG_OBJECT(boilerplatevideofilter, "set_property");

  switch (property_id) {
    case PROP_INT_PROPERTY:
      boilerplatevideofilter->int_property = g_value_get_int(value);
      break;

    case PROP_BOOL_PROPERTY:
      boilerplatevideofilter->bool_property = g_value_get_boolean(value);
      break;

    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
  }
}

void
gst_boilerplatevideofilter_get_property (GObject * object, guint property_id,
    GValue * value, GParamSpec * pspec)
{
  GstBoilerPlateVideoFilter *boilerplatevideofilter = GST_BOILERPLATE_VIDEOFILTER (object);

  GST_DEBUG_OBJECT (boilerplatevideofilter, "get_property");

  switch (property_id) {
    case PROP_INT_PROPERTY:
      g_value_set_int(value, boilerplatevideofilter->int_property);
      break;
      
    case PROP_BOOL_PROPERTY:
      g_value_set_boolean(value, boilerplatevideofilter->bool_property);
      break;

    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
  }
}

static gboolean
gst_boilerplatevideofilter_set_info (GstVideoFilter * filter, GstCaps * incaps,
    GstVideoInfo * in_info, GstCaps * outcaps, GstVideoInfo * out_info)
{
  GstBoilerPlateVideoFilter *boilerplatevideofilter = GST_BOILERPLATE_VIDEOFILTER (filter);

  GST_DEBUG_OBJECT (boilerplatevideofilter, "set_info");

  return TRUE;
}

/* transform */
static GstFlowReturn
gst_boilerplatevideofilter_transform_frame (GstVideoFilter * filter,
    GstVideoFrame * inframe, GstVideoFrame * outframe)
{
  GstBoilerPlateVideoFilter *boilerplatevideofilter = GST_BOILERPLATE_VIDEOFILTER (filter);

  GST_DEBUG_OBJECT (boilerplatevideofilter, "transform_frame");

  return GST_FLOW_OK;
}

static GstFlowReturn
gst_boilerplatevideofilter_transform_frame_ip (GstVideoFilter * filter,
    GstVideoFrame * frame)
{
  GstBoilerPlateVideoFilter *boilerplatevideofilter = GST_BOILERPLATE_VIDEOFILTER (filter);

  return GST_FLOW_OK;
}

static gboolean
plugin_init (GstPlugin * plugin)
{

  /* FIXME Remember to set the rank if it's an element that is meant
     to be autoplugged by decodebin. */
  return gst_element_register (plugin, "boilerplatevideofilter", GST_RANK_NONE,
      GST_TYPE_BOILERPLATE_VIDEOFILTER);
}

/* FIXME: these are normally defined by the GStreamer build system.
   If you are creating an element to be included in gst-plugins-*,
   remove these, as they're always defined.  Otherwise, edit as
   appropriate for your external plugin package. */
#ifndef VERSION
#define VERSION THISBUILD
#endif
#ifndef PACKAGE
#define PACKAGE "gst_boilerplate_videofilter_package"
#endif
#ifndef PACKAGE_NAME
#define PACKAGE_NAME "gst_boilerplate_videofilter"
#endif
#ifndef GST_PACKAGE_ORIGIN
#define GST_PACKAGE_ORIGIN "https://github.com/imrisaac/cmake-gst-plugin-template"
#endif

GST_PLUGIN_DEFINE (GST_VERSION_MAJOR,
    GST_VERSION_MINOR,
    boilerplatevideofilter,
    "Boilerplate template for gstreamer videofilter plugin using cmake build system",
    plugin_init, VERSION, "LGPL", PACKAGE_NAME, GST_PACKAGE_ORIGIN)
